### PR TITLE
Make upgrade work on global installs

### PR DESCRIPTION
**Pull request by amcaplan**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/shopify-cli-planning/issues/378

We have an upgrade command that works on per-project installs, but not on system-wide installs.

I realized we will nearly always do `npm install -g` to upgrade, even if we end up supporting Linux with explicit packages, and the only real exception is Homebrew. So may as well just have this command work in global installs.

### WHAT is this pull request doing?

* Checks whether we're running the CLI inside a package manager. If yes, proceed as before.
* If not, apparently we're running the CLI globally and we should upgrade globally.
* Check if it's been installed via Homebrew, and install the appropriate formula.
* If it's not installed via Homebrew, install via `npm install -g`.

I also added significant testing around the whole flow.

Note that I could have opted for a service object, but for now I wanted to avoid introducing that concept to cli-main, especially since this is a specialized workflow that isn't likely to change.

### How to test your changes?

```sh
$ bin/run.js upgrade --path=/Users/YOURUSER
```

Try with/without Homebrew-installed CLI, with an installed older version, etc.

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [x] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes

### CODE REVIEW

File: packages/cli-main/src/cli/commands/upgrade.ts [hunk 1]
```diff
@@ -21,43 +22,27 @@ export default class Upgrade extends Command {
   }
 
   async run(): Promise<void> {
```
Comment (pepicrft) [inline packages/cli-main/src/cli/commands/upgrade.ts:24]: I'd advise extracting all the business logic into a service to allow unit-testing instead of integration-testing through oclif's primitives.

File: packages/cli-main/src/cli/commands/upgrade.ts [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
-import {error, file, output, path} from '@shopify/cli-kit'
+import {error, file, os, output, path, system} from '@shopify/cli-kit'
 import {
   addNPMDependencies,
   checkForNewVersion,
   DependencyType,
   getPackageManager,
 } from '@shopify/cli-kit/node/node-package-manager'
 import Command from '@shopify/cli-kit/node/base-command'
+const cliDependency = '@shopify/cli'
```
Comment (pepicrft) [inline packages/cli-main/src/cli/commands/upgrade.ts:10]: Considering this information is in the `package.json` of the package containing this module, I'd keep that as the source of truth and read the information at runtime.

File: packages/cli-kit/src/version.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import {content, token, debug} from './output.js'
-import {moduleDirectory} from './path.js'
+import {findUp, moduleDirectory} from './path.js'
```
Comment (pepicrft) [inline packages/cli-kit/src/version.ts:2]: Is this change needed?

File: packages/cli-main/src/cli/commands/upgrade.test.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  })
+  vi.mocked(os.platformAndArch).mockReturnValue({platform: 'win32', arch: 'amd64'})
+})
+afterEach(() => {
+  outputMocker.mockAndCaptureOutput().clear()
+})
+
+describe('upgrade global CLI', () => {
+  it('does not upgrade globally if the latest version is found', async () => {
+    const currentCliVersion = await constants.versions.cliKit()
```
Comment (pepicrft) [inline packages/cli-main/src/cli/commands/upgrade.test.ts:36]: Why not moving this one inside the `inTemporaryDirectory`?

File: packages/cli-main/src/cli/services/upgrade.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  findUpAndReadPackageJson,
+  checkForNewVersion,
+  DependencyType,
+  getPackageManager,
+} from '@shopify/cli-kit/node/node-package-manager'
+
+// Canonical list of oclif plugins that should be installed globally
+const globalPlugins = ['@shopify/theme']
+
+interface PackageJsonContents {
```
Comment (pepicrft) [inline packages/cli-main/src/cli/services/upgrade.ts:13]: What about reusing [this model](https://github.com/Shopify/cli/blob/main/packages/cli-kit/src/node/node-package-manager.ts#L210) from `@shopify/cli-kit`?

File: packages/cli-main/src/cli/services/upgrade.ts [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  }
+}
+
+export async function upgrade(directory: string, currentVersion: string): Promise<void> {
+  let newestVersion: string | void
+
+  const projectDir = await getProjectDir(directory)
+  if (projectDir) {
+    newestVersion = await upgradeLocalShopify(projectDir, currentVersion)
+  } else if (usingPackageManager()) {
```
Comment (pepicrft) [inline packages/cli-main/src/cli/services/upgrade.ts:28]: I find this clause a bit confusing. What does it mean that the `npm_config_user_agent` environment variable is not present? Who is setting that variable and when?

File: packages/cli-main/src/cli/services/upgrade.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+async function upgradeLocalShopify(projectDir: string, currentVersion: string): Promise<string | void> {
+  const packageJson = (await findUpAndReadPackageJson(projectDir)).content as PackageJsonContents
+  const packageJsonDependencies = packageJson.dependencies || {}
+  const packageJsonDevDependencies = packageJson.devDependencies || {}
+
+  let resolvedVersion: string = {...packageJsonDependencies, ...packageJsonDevDependencies}[await cliDependency()]!
+  if (resolvedVersion.slice(0, 1).match(/[\^~]/)) resolvedVersion = currentVersion
+  const newestVersion = await checkForNewVersion(await cliDependency(), resolvedVersion)
+
+  if (!newestVersion) return wontInstall(resolvedVersion)
```
Comment (pepicrft) [inline packages/cli-main/src/cli/services/upgrade.ts:69]: `wontInstall` doesn't convey what the method does. What about:

```js
if (!newestVersion) {
  outputWontInstallMessage(resolvedVersion)
  return;
}

File: packages/cli-main/src/cli/services/upgrade.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+  outputUpgradeMessage(resolvedVersion, newestVersion)
+
+  await installJsonDependencies('prod', packageJsonDependencies, projectDir)
```
Comment (pepicrft) [inline packages/cli-main/src/cli/services/upgrade.ts:73]: If we pass `packageJsonDependencies` aren't we updating all the dependencies existing in the project?

File: packages/cli-main/src/cli/services/upgrade.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+      // eslint-disable-next-line no-catch-all/no-catch-all
+    } catch (err) {
+      output.warn('Homebrew upgrade failed. Falling back to standard npm install.')
+    }
+  }
+  const command = 'npm'
+  const args = [
+    'install',
+    '-g',
+    `${await cliDependency()}@latest`,
```
Comment (pepicrft) [inline packages/cli-main/src/cli/services/upgrade.ts:107]: I see more than one `await cliDependency()` scattered in this function. What about calling it once and passing it down?

File: packages/cli-main/src/cli/services/upgrade.ts (hunk 6 above)
Comment (amcaplan) [reply packages/cli-main/src/cli/services/upgrade.ts:28]: It's a [standard variable](https://docs.npmjs.com/cli/v8/using-npm/config#user-agent) set by the package managers. If it's there, we assume this is in a project and therefore the upgrade is local. Otherwise we're assuming a global upgrade.

File: packages/cli-main/src/cli/services/upgrade.ts (hunk 8 above)
Comment (amcaplan) [reply packages/cli-main/src/cli/services/upgrade.ts:73]: No, since `installJsonDependencies filters to just `@shopify/cli` and canonical oclif plugins.

File: packages/cli-main/src/cli/services/upgrade.ts [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        )
+        await system.exec('brew', ['upgrade', 'shopify-cli@3'], {stdio: 'inherit'})
+        return newestVersion
+      }
+      // eslint-disable-next-line no-catch-all/no-catch-all
+    } catch (err) {
+      output.warn('Homebrew upgrade failed. Falling back to standard npm install.')
+    }
+  }
+  const command = 'npm'
```
[the remark this answers is not in the document]
Comment (amcaplan) [reply packages/cli-main/src/cli/services/upgrade.ts:103]: Good point. Also if it's MacOS but it was installed by `npm`, this will have weird output.

File: packages/cli-main/src/cli/services/upgrade.ts [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+          return file.exists(path.join(dir, configFile))
+        }),
+      )
+      if (configFilesExist.some((bool) => bool)) return dir
+    },
+    {
+      cwd: directory,
+      type: 'directory',
+    },
+  )
```
[the remark this answers is not in the document]
Comment (amcaplan) [reply packages/cli-main/src/cli/services/upgrade.ts:57]: This was a hack because findUp doesn't document supporting an array of paths (as far as I could tell). However, you're right that [it does](https://github.com/sindresorhus/find-up/blob/a3b2f1e788e3c4023eab8f6282b085b57131ae15/index.js#L14). Way better!

**Review by pepicrft — APPROVED** [review]
